pkg/ring/tokens.go

```diff
@@ -0,0 +1,95 @@
+package ring
+
+import (
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"os"
+
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+const (
+	// TokenVersion1 is the version is a simple list of tokens.
+	TokenVersion1 = 1
+)
+
+// Tokens is a simple list of tokens.
+type Tokens []uint32
+
+// StoreToFile stores the tokens in the given directory.
+func (t Tokens) StoreToFile(tokenFilePath string) error {
+	if tokenFilePath == "" {
+		return errors.New("path is empty")
+	}
+	tmp := tokenFilePath + ".tmp"
```

Comment:
This is not safe if `StoreToFile()` is called concurrently (we will have two go routines writing to the same `.tmp` file). You should use https://golang.org/pkg/io/ioutil/#TempFile

Reply:
If we are doing https://github.com/cortexproject/cortex/pull/1750#discussion_r346005496, it fixes this. Else will use the random temp file.

Reply:
I wouldn't advocate to wrap the `StoreToFile()` within the lock. I would personally suggest to use `ioutil.TempFile()` here.